# Day 4 EOD: Microplastics, From Raw File to Real Numbers

Date: 09/03/2026

In [31]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

url = 'https://eds-217-essential-python.github.io/data/marine_microplastics.csv'
df = pd.read_csv(url, parse_dates=['Date'], date_format='%m/%d/%Y %I:%M:%S %p')

In [32]:
df['Unit'].value_counts()

Unit
pieces/m3           10178
pieces/10 mins       5792
pieces kg-1 d.w.      275
Name: count, dtype: int64

In [33]:
df

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
0,10008,Atlantic Ocean,NaN,NaN,Grab sample,0.020000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-58.428300,-64.164000,2017-02-03,1e5b8e71-037b-4887-a276-f1e4552acb1f,-64.164000,-58.428300
1,8680,Atlantic Ocean,NaN,NaN,Grab sample,0.008000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-51.308200,-60.546700,2013-11-17,a40f7f7c-1025-4aac-ad16-ee4cba196870,-60.546700,-51.308200
2,13257,Pacific Ocean,NaN,NaN,Manta net,0.019886,pieces/m3,0.005-1,Medium,Faure et al.2015,...,"Oceaneye Association, Switzerland",Oceaneye Association; Citizen Science,276422,https://www.ncei.noaa.gov/access/metadata/land...,-51.826667,-72.575000,2015-12-26,febf79b8-7e2c-46e6-bc15-e08492ec2029,-72.575000,-51.826667
3,9676,Atlantic Ocean,NaN,NaN,Grab sample,0.018000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-31.696000,-48.560000,2015-08-11,a77121b2-e113-444e-82d9-7af11d62fdd2,-48.560000,-31.696000
4,6427,Pacific Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2014,...,Sea Education Association,SEA,211008,https://www.ncei.noaa.gov/access/metadata/land...,6.350000,-121.850000,2002-12-18,be27c450-02ca-4261-8d89-cae21108e6cc,-121.850000,6.350000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16240,10726,Atlantic Ocean,Mediterranean Sea,Balearic Sea,Manta net,0.326000,pieces/m3,0.005-1,Medium,de Haan et al.2022,...,University of Barcelona,Citizen science; Surf board; Kayak,253923,https://www.ncei.noaa.gov/access/metadata/land...,40.910469,0.833376,2021-03-11,b840a79b-6d98-4def-ac53-5447be1531b9,0.833376,40.910469
16241,9366,Atlantic Ocean,NaN,NaN,Grab sample,0.011000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,17.100000,-47.550000,2015-02-01,fd4db6d6-aed7-48b1-a22b-402557112e2f,-47.550000,17.100000
16242,4683,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.021600,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,21.390000,-81.960000,2001-03-14,83a8208e-2d06-4ad9-bcd9-81e29d57e0aa,-81.960000,21.390000
16243,6043,Atlantic Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,24.140000,-81.980000,2008-05-01,0fd72d83-cd27-42e1-b63a-b015c23b28d2,-81.980000,24.140000


16,245 rows and 22 columns

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16245 entries, 0 to 16244
Data columns (total 22 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   OBJECTID          16245 non-null  int64         
 1   Oceans            15974 non-null  object        
 2   Regions           7996 non-null   object        
 3   SubRegions        588 non-null    object        
 4   Sampling Method   16245 non-null  object        
 5   Measurement       10453 non-null  float64       
 6   Unit              16245 non-null  object        
 7   Density Range     16245 non-null  object        
 8   Density Class     16245 non-null  object        
 9   Short Reference   16245 non-null  object        
 10  Long Reference    16245 non-null  object        
 11  DOI               16245 non-null  object        
 12  Organization      16245 non-null  object        
 13  Keywords          16227 non-null  object        
 14  Accession Number  1624

Which columns are object, and which of those should be something else?

Oceans, regions, subregions, sampling method, unit, density range, density class, short reference, long reference, DOI, organization, keywords, accession link, and globalID are all objects.

All of these should be objects

In [35]:
df.isnull().sum()

OBJECTID                0
Oceans                271
Regions              8249
SubRegions          15657
Sampling Method         0
Measurement          5792
Unit                    0
Density Range           0
Density Class           0
Short Reference         0
Long Reference          0
DOI                     0
Organization            0
Keywords               18
Accession Number        0
Accession Link          0
Latitude                0
Longitude               0
Date                    0
GlobalID                0
x                       0
y                       0
dtype: int64

How many columns have gaps? Which are the columns, and how many gaps do they have?

oceans, region, subregion, measurement, and keyword all have missing data.  On rows where oceans is missing, region is also missing.

In [36]:
df = df.dropna(subset=['Measurement'])
df

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
0,10008,Atlantic Ocean,NaN,NaN,Grab sample,0.020000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-58.428300,-64.164000,2017-02-03,1e5b8e71-037b-4887-a276-f1e4552acb1f,-64.164000,-58.428300
1,8680,Atlantic Ocean,NaN,NaN,Grab sample,0.008000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-51.308200,-60.546700,2013-11-17,a40f7f7c-1025-4aac-ad16-ee4cba196870,-60.546700,-51.308200
2,13257,Pacific Ocean,NaN,NaN,Manta net,0.019886,pieces/m3,0.005-1,Medium,Faure et al.2015,...,"Oceaneye Association, Switzerland",Oceaneye Association; Citizen Science,276422,https://www.ncei.noaa.gov/access/metadata/land...,-51.826667,-72.575000,2015-12-26,febf79b8-7e2c-46e6-bc15-e08492ec2029,-72.575000,-51.826667
3,9676,Atlantic Ocean,NaN,NaN,Grab sample,0.018000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-31.696000,-48.560000,2015-08-11,a77121b2-e113-444e-82d9-7af11d62fdd2,-48.560000,-31.696000
4,6427,Pacific Ocean,NaN,NaN,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2014,...,Sea Education Association,SEA,211008,https://www.ncei.noaa.gov/access/metadata/land...,6.350000,-121.850000,2002-12-18,be27c450-02ca-4261-8d89-cae21108e6cc,-121.850000,6.350000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16239,8917,Atlantic Ocean,NaN,NaN,Grab sample,0.004000,pieces/m3,0.0005-0.005,Low,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,22.341300,-19.952500,2014-11-12,fa761396-9019-4c6a-a7e5-a4822a8f2926,-19.952500,22.341300
16240,10726,Atlantic Ocean,Mediterranean Sea,Balearic Sea,Manta net,0.326000,pieces/m3,0.005-1,Medium,de Haan et al.2022,...,University of Barcelona,Citizen science; Surf board; Kayak,253923,https://www.ncei.noaa.gov/access/metadata/land...,40.910469,0.833376,2021-03-11,b840a79b-6d98-4def-ac53-5447be1531b9,0.833376,40.910469
16241,9366,Atlantic Ocean,NaN,NaN,Grab sample,0.011000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,17.100000,-47.550000,2015-02-01,fd4db6d6-aed7-48b1-a22b-402557112e2f,-47.550000,17.100000
16242,4683,Atlantic Ocean,Caribbean Sea,NaN,Neuston net,0.021600,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,21.390000,-81.960000,2001-03-14,83a8208e-2d06-4ad9-bcd9-81e29d57e0aa,-81.960000,21.390000


5792 rows were dropped

In [37]:
df['Unit'].value_counts()

Unit
pieces/m3           10178
pieces kg-1 d.w.      275
Name: count, dtype: int64

The "pieces/10 mins: 5,792" row disappeared.  This is because our dropna dropped all 5792 rows with that sampling method. 

In [44]:
# Fill the gaps in Regions and SubRegions with 'Unspecified', one column at a time, 
# assigning each result back. 
# Confirm with .isnull().sum() that only Keywords is still missing anything

df['Regions'] = df['Regions'].fillna('Unspecified')
df['SubRegions'] = df['SubRegions'].fillna('Unspecified')
df.isnull().sum()

<positron-console-cell-44>:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
<positron-console-cell-44>:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


OBJECTID             0
Oceans               0
Regions              0
SubRegions           0
Sampling Method      0
Measurement          0
Unit                 0
Density Range        0
Density Class        0
Short Reference      0
Long Reference       0
DOI                  0
Organization         0
Keywords            18
Accession Number     0
Accession Link       0
Latitude             0
Longitude            0
Date                 0
GlobalID             0
x                    0
y                    0
dtype: int64

In [46]:
# Check for duplicated rows. 
# Do it twice: once bare, and once with subset= naming the columns that define a 
# single observation, which here are Latitude, Longitude, Date, Measurement and Unit. 
# Report both numbers and say which one you believe, and why the two differ so much.

print(df.duplicated().sum())
print(df.duplicated(subset=['Latitude', 'Longitude', 'Date', 'Measurement', 'Unit']).sum())

0
539


df.duplicated said there were zero duplicates. Adding a subset changed the number of duplicates rows to 539.  The difference is probably because there are no rows where ALL columns are the same, but there are many rows where just the 5 subsetted columns are the same.  Which value is "correct" depends on the purpose.  It is likely that some of the subsetted "duplicates" could not actually be duplicates if multiple measurements are collected from the same site on the same day.

In [48]:
samples = df[df['Unit'] == 'pieces/m3'].copy()  
samples

,OBJECTID,Oceans,Regions,SubRegions,Sampling Method,Measurement,Unit,Density Range,Density Class,Short Reference,...,Organization,Keywords,Accession Number,Accession Link,Latitude,Longitude,Date,GlobalID,x,y
0,10008,Atlantic Ocean,Unspecified,Unspecified,Grab sample,0.020000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-58.428300,-64.164000,2017-02-03,1e5b8e71-037b-4887-a276-f1e4552acb1f,-64.164000,-58.428300
1,8680,Atlantic Ocean,Unspecified,Unspecified,Grab sample,0.008000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-51.308200,-60.546700,2013-11-17,a40f7f7c-1025-4aac-ad16-ee4cba196870,-60.546700,-51.308200
2,13257,Pacific Ocean,Unspecified,Unspecified,Manta net,0.019886,pieces/m3,0.005-1,Medium,Faure et al.2015,...,"Oceaneye Association, Switzerland",Oceaneye Association; Citizen Science,276422,https://www.ncei.noaa.gov/access/metadata/land...,-51.826667,-72.575000,2015-12-26,febf79b8-7e2c-46e6-bc15-e08492ec2029,-72.575000,-51.826667
3,9676,Atlantic Ocean,Unspecified,Unspecified,Grab sample,0.018000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,-31.696000,-48.560000,2015-08-11,a77121b2-e113-444e-82d9-7af11d62fdd2,-48.560000,-31.696000
4,6427,Pacific Ocean,Unspecified,Unspecified,Neuston net,0.000000,pieces/m3,0-0.0005,Very Low,Law et al.2014,...,Sea Education Association,SEA,211008,https://www.ncei.noaa.gov/access/metadata/land...,6.350000,-121.850000,2002-12-18,be27c450-02ca-4261-8d89-cae21108e6cc,-121.850000,6.350000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16239,8917,Atlantic Ocean,Unspecified,Unspecified,Grab sample,0.004000,pieces/m3,0.0005-0.005,Low,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,22.341300,-19.952500,2014-11-12,fa761396-9019-4c6a-a7e5-a4822a8f2926,-19.952500,22.341300
16240,10726,Atlantic Ocean,Mediterranean Sea,Balearic Sea,Manta net,0.326000,pieces/m3,0.005-1,Medium,de Haan et al.2022,...,University of Barcelona,Citizen science; Surf board; Kayak,253923,https://www.ncei.noaa.gov/access/metadata/land...,40.910469,0.833376,2021-03-11,b840a79b-6d98-4def-ac53-5447be1531b9,0.833376,40.910469
16241,9366,Atlantic Ocean,Unspecified,Unspecified,Grab sample,0.011000,pieces/m3,0.005-1,Medium,Barrows et al.2018,...,Adventure Scientist,Adventure Scientist/Citizen Science,211009,https://www.ncei.noaa.gov/access/metadata/land...,17.100000,-47.550000,2015-02-01,fd4db6d6-aed7-48b1-a22b-402557112e2f,-47.550000,17.100000
16242,4683,Atlantic Ocean,Caribbean Sea,Unspecified,Neuston net,0.021600,pieces/m3,0.005-1,Medium,Law et al.2010,...,Sea Education Association,SEA,211007,https://www.ncei.noaa.gov/access/metadata/land...,21.390000,-81.960000,2001-03-14,83a8208e-2d06-4ad9-bcd9-81e29d57e0aa,-81.960000,21.390000


In [51]:
df['Accession Number'] = df['Accession Number'].astype(str)
samples['Accession Number'] = samples['Accession Number'].astype(str)

<positron-console-cell-51>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [52]:
samples['Accession Number'].dtype

dtype('O')